### PR TITLE
Addd mkcert and make chrome trust https URLs, fixes #15

### PR DESCRIPTION
**Pull request by rfay**
#15 points out that https URLs aren't trusted by chrome here.

This installs mkcert in the chrome container and makes it trust the shared mkcert CA.
curl also works inside the chrome container.

You can test this PR with 

`ddev get https://github.com/rfay/ddev-selenium-standalone-chrome/tarball/20230121_trust_certs`

This seems to demonstrate that the [demonstration repo](https://github.com/ptmkenny/d10-behat) provided by @ptmkenny in #15 will work OK, trusting https URLs.

### CODE REVIEW

File: selenium-standalone-chrome/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#ddev-generated
+
+FROM seleniarm/standalone-chromium:4.1.4-20220429
+ENV CAROOT=/mnt/ddev-global-cache/mkcert
+
+USER root
+RUN apt update >/dev/null && apt install -y iputils-ping less mkcert vim >/dev/null
```
Comment (rfay) [inline selenium-standalone-chrome/Dockerfile:7]: I guess this could be done with a curl install in post-start hook, but it's less intrusive here. But that would get rid of the whole Dockerfile.

File: docker-compose.selenium-chrome.yaml [hunk 2]
```diff
@@ -7,7 +7,7 @@
 version: '3.6'
 services:
   selenium-chrome:
-    image: seleniarm/standalone-chromium:4.1.4-20220429
+    build: selenium-standalone-chrome
```
Comment (rfay) [inline docker-compose.selenium-chrome.yaml:10]: This just says build using the Dockerfile in the selenium-standalone-chrome directory.

File: docker-compose.selenium-chrome.yaml [hunk 3]
```diff
@@ -27,7 +27,8 @@ services:
       com.ddev.approot: $DDEV_APPROOT
     volumes:
       - ".:/mnt/ddev_config"
+      - ddev-global-cache:/mnt/ddev-global-cache
```
Comment (rfay) [inline docker-compose.selenium-chrome.yaml:30]: This volume has the mkcert CA in it.

File: config.selenium-standalone-chrome.yaml [hunk 4]
```diff
@@ -4,6 +4,16 @@
 #
 # This file comes from https://github.com/drud/ddev-selenium-standalone-chrome
 #
+hooks:
+  # Make sure that the standard mkcert CA keys are readable
+  # as they normally are owned by the ddev-defined user, and here we're running
+  # as 'seluser'
+  # and do a mkcert -install so we can trust them
+  post-start:
+    - exec: "sudo chmod ugo+r /mnt/ddev-global-cache/mkcert/*"
```
Comment (rfay) [inline config.selenium-standalone-chrome.yaml:13]: DDEV has installed the mkcert CA for use by its installed user (based on the host-side user and userid) but the chrome container is running as user `seluser`, so we need to make it readable. This sudo chmod takes place using the web container, which has sudo in it.

File: config.selenium-standalone-chrome.yaml [hunk 5]
```diff
@@ -4,6 +4,16 @@
...
+    - exec: "mkcert -install"
+      service: selenium-chrome
```
Comment (rfay) [inline config.selenium-standalone-chrome.yaml:15]: And here we do the mkcert -install inside the selenium-chrome container. This can't be done earlier (like in Dockerfile) because we have to have access to the CA, mounted in /mnt/ddev-global-cache/mkcert

**Comment by ptmkenny** [comment]
So I tried running `ddev get https://github.com/rfay/ddev-selenium-standalone-chrome/tarball/20230121_trust_certs` in the test repo and then doing `ddev restart` but I couldn't get it working-- when I run `ddev composer behat`, I'm seeing an SSL failure after removing the `--ignore-certificates` flag in `behat-ddev-selenium.yml`.

I then tried to delete the data and start over with `ddev delete` and then `ddev drush site:install --account-name=admin --account-pass=admin -y` but I got the same error.

I committed my changes to a new branch so maybe you can see what I'm doing wrong?

https://github.com/ptmkenny/d10-behat/compare/main...pr17

**Comment by rfay** [comment]
* I'm frustrated that you had the `--ignore-certificate-errors` in your repo that was supposed to demonstrate the error. Maybe I missed something along the way.
* And I guess I don't know why you now care to remove that, and what trouble it causes. 

I didn't have any trouble with your repo and `ddev composer behat` as you had put it out there to demo. Seems like you're changing the rules of the game :)

I imagine headless chrome needs something to install the CA at a different point, not sure. 

Can you please get your repo to a place where it demonstrates the problem we're trying to solve? Or is that what the pr17 branch now is?

**Comment by rfay** [comment]
Please also help me to understand how to know that the failure which I now see is a result of untrusted cert. Behat has always been a bit inscrutable in its explanation of what's going on. How can I see the cert failure?

**Comment by rfay** [comment]
Ah, I see I can see the interaction at 
`https://d10-behat.ddev.site:7900/`

**Comment by ptmkenny** [comment]
I'm sorry to cause you frustration.  I'm going to restate the problem to avoid further miscommunication and to make sure that we have the same understanding.

The issue at hand is that when testing with behat using the selenium-standalone-chrome container, you get a certificate error if you test with an https URL. This error can be worked around by setting the  `--ignore-certificate-errors` flag for Chrome in the behat config file (in my repo's case, `tests/behat/behat-ddev-selenium.yml`.  Since this allows behat tests to be run, I submitted a PR that added a note to the README about setting this flag.  Then you suggested it would be better to fix this in the container itself if possible by getting it to trust the certificate (which I agree- that would definitely be better because it's zero config on the part of the user).

The [repo](https://github.com/ptmkenny/d10-behat) I added was a "working" demonstration of an https URL with the `--ignore-certificates` flag.  That's probably the source of the confusion-- I submitted a repo in a working state instead of in a broken state because (to my mind) I was demonstrating the "fix" (my workaround) rather than a bug that needed to be fixed, but thinking about it now it makes more sense to submit a broken repo showing the problem and then an additional branch showing the fix, so I'm sorry about that.

So, to sum up, to see the problem, `--ignore-certificate-errors` should not be set (as is the case in pr-17, so pr-17 is demonstrating the problem, whereas the original `main` branch is "fixing" the https issue by setting `--ignore-certificate-errors`).

**Comment by rfay** [comment]
Looks like the installation of CA is covered in https://github.com/seleniumhq-community/docker-seleniarm#install-certificates-for-chromium-based-browsers, and apparently it doesn't trust the system CA by default.

**Comment by rfay** [comment]
But @ptmkenny does `--ignore-certificate-errors` solve all your problems? Is that the easy path?

**Comment by ptmkenny** [comment]
> But @ptmkenny does `--ignore-certificate-errors` solve all your problems? Is that the easy path?

Yes, it solves all my problems, specifically:

* All the tests run as expected.  All https URLs, including images and CSS, load properly in the tests.
* http to https redirects on the test site work.

I think this is also a pretty easy way to do it because if you are running behat tests like this you probably already have a `behat.yml`, and this is just adding one line to that file.

**Comment by rfay** [comment]
The chrome launched in the container should be able to accept `--ignore-certificate-errors` by default, but I haven't figured out how to get it to do that, despite a number of resources.
* https://stackoverflow.com/a/25063706/215713
* https://stackoverflow.com/a/49835651/215713
* https://sahajamit.medium.com/selenium-chrome-dev-tools-makes-a-perfect-browser-automation-recipe-c35c7f6a2360
* https://stackoverflow.com/a/74733865/215713
* http://blogs.stevelongchen.com/2020/05/14/selenium-how-to-disable-or-bypass-connection-is-not-private-for-chrome-and-firefox/

**Comment by rfay** [comment]
I would have thought that just this would do the job,
`  - MINK_DRIVER_ARGS_WEBDRIVER=[\"chrome\", {\"browserName\":\"chrome\",\"chromeOptions\":{\"w3c\":false,\"args\":[\"--disable-gpu\",\"--headless\", \"--no-sandbox\", \"--disable-dev-shm-usage\", \"--ignore-certificate-errors\",\"--allow-insecure-localhost\"]}}, \"http://selenium-chrome:4444/wd/hub\"]`

**Comment by rfay** [comment]
I'm pretty sure we can make the selenium-chrome container trust the mkcert CA with [docs](https://github.com/seleniumhq-community/docker-seleniarm#install-certificates-for-chromium-based-browsers), but it would be so very much easier to just figure out where the magic goes for defaulting `--ignore-certificate-errors`

**Comment by ptmkenny** [comment]
> I would have thought that just this would do the job, ` - MINK_DRIVER_ARGS_WEBDRIVER=[\"chrome\", {\"browserName\":\"chrome\",\"chromeOptions\":{\"w3c\":false,\"args\":[\"--disable-gpu\",\"--headless\", \"--no-sandbox\", \"--disable-dev-shm-usage\", \"--ignore-certificate-errors\",\"--allow-insecure-localhost\"]}}, \"http://selenium-chrome:4444/wd/hub\"]`

I am not an expert on this, but fixing it here should work for anyone using the default args, so the only people who wouldn't be covered by this case would (I think) be people like me who are specifying additional chrome options/setting up a browser profile in a behat config.yaml, which will require setting the flag manually.

**Comment by rfay** [comment]
@ptmkenny that's what I was aiming for. All the changes discussed here were aiming at that. The quote I gave was something in the docker-compose.

**Review by weitzman — APPROVED** [review]

**Comment by rfay** [comment]
Closing this for now. Happy to return to it later or work on solving this problem another way. For now it's distracting for it to be in here.
